envoy/stats/allocator.h

```diff
@@ -58,6 +58,25 @@ class Allocator {
   virtual const SymbolTable& constSymbolTable() const PURE;
   virtual SymbolTable& symbolTable() PURE;
 
+  /**
+   * Iterate over all stats that need to be sinked. Note, that implementations can potentially hold
+   * on to a mutex that will deadlock if the passed in functors try to create or delete a stat.
+   * @param f_size functor that is provided the number of all sinked stats.
+   * @param f_stat functor that is provided one sinked stat at a time.
+   */
+  virtual void forEachSinkedCounter(std::function<void(std::size_t)> f_size,
+                                    std::function<void(Stats::Counter&)> f_stat) PURE;
+  virtual void forEachSinkedGauge(std::function<void(std::size_t)> f_size,
+                                  std::function<void(Stats::Gauge&)> f_stat) PURE;
+  virtual void forEachSinkedTextReadout(std::function<void(std::size_t)> f_size,
+                                        std::function<void(Stats::TextReadout&)> f_stat) PURE;
+
+  /**
+   * @param filter should return true if the passed in stat needs to be sinked.
+   */
+  virtual void setCounterSinkFilter(std::function<bool(const Stats::Counter&)> filter) PURE;
```

Comment:
Consider splitting this into 2 PRs. You have a pure implementation improvement to allow iterating over the sinked stats without building an intermediate array, which is strictly a win and we should check that in as an intermediate point. However I don't think it's going to be enough based on the flame graphs.

You also have an implementation of an idea we came up with, but no users of it yet. I think we should use a doc & meeting to get buy-in for it internally and in the Envoy community. And for the community we probably need a layered solution that doesn't require a custom C++ call, e.g. using StatsMatcher or similar. Maybe others that have been bit by stats-sink scalability will have ideas on how to make the filters usable without hacking the binary.

While we iterate toward a solution that will benefit everyone, let's get the easy win you already found by switching to the iterator.